src/tests/async/returns.cs

```diff
@@ -34,11 +34,19 @@ private static async2 Task Returns(C c)
             AssertEqual(424242, c.Val.C);
             AssertEqual(42424242, c.Val.D);
 
-            S<string> strings = await ReturnsStructGC();
-            AssertEqual("A", strings.A);
-            AssertEqual("B", strings.B);
-            AssertEqual("C", strings.C);
-            AssertEqual("D", strings.D);
+            // TODO: need to fix this
```

Comment:
@jakobbotsch the change stresses calling via thunks and possibly introduced some scenarios that tests did not cover before. Remarkably, nearly everything works fine!! However, here I saw an assert and turned off one scenario. 
Not sure if this is something wrong with IL or something on the JIT side.  
(the other case is with thunks for async methods in structs).

Reply:
The failure does not happen with optimization enabled, so it is not caused by that. 
(i.e. by default this test passes, but with `set DOTNET_JitOptimizeAwait=0` I see a failure). 

I'll try to take a look.

Reply:
It looks like the async resumption stub in this case needs to box the result, since the result is an obj-containing struct.  

We are making a resumption stub for  `RuntimeHelpers.Await<T>(Task<T>)` when `T` is `S<string>`, thus the sig reports the result type as `S'1<System.__Canon>` and we emit Box instruction with that.

Reply:
I think this is a general issue with shared generics and nongeneric resumption stub that may need to box the result. 
I can hit this in shared-generics.cs test by adding scenario like:
(regardless of Await optimization)

```cs
        Async1EntryPoint<S<string>>(typeof(S<string>), new S<string> {t = "ghj" }).Wait();
```

not sure yet how to deal with this. 

It does not seem like the fix would be in the scope of this PR.

Reply:
So can this be uncommented then?

Can you also open an issue about the unhandled case?

Reply:
logged https://github.com/dotnet/runtimelab/issues/3010